src/nvim/eval/typval.c

```diff
@@ -0,0 +1,2556 @@
+{
+  list_T *const list = xcalloc(1, sizeof(list_T));
+
+  // Prepend the list to the list of lists for garbage collection.
+  if (gc_first_list != NULL) {
+    gc_first_list->lv_used_prev = list;
+  }
+  list->lv_used_prev = NULL;
+  list->lv_used_next = gc_first_list;
+  gc_first_list = list;
+  return list;
+}
+
+/// Free items contained in a list
+///
+/// @param[in,out]  l  List to clear.
+void tv_list_free_contents(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  for (listitem_T *item = l->lv_first; item != NULL; item = l->lv_first) {
+    // Remove the item before deleting it.
+    l->lv_first = item->li_next;
+    tv_clear(&item->li_tv);
+    xfree(item);
+  }
+  l->lv_len = 0;
+  l->lv_idx_item = NULL;
+  l->lv_last = NULL;
+  assert(l->lv_watch == NULL);
+}
+
+/// Free a list itself, ignoring items it contains
+///
+/// Ignores the reference count.
+///
+/// @param[in,out]  l  List to free.
+void tv_list_free_list(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  // Remove the list from the list of lists for garbage collection.
+  if (l->lv_used_prev == NULL) {
+    gc_first_list = l->lv_used_next;
+  } else {
+    l->lv_used_prev->lv_used_next = l->lv_used_next;
+  }
+  if (l->lv_used_next != NULL) {
+    l->lv_used_next->lv_used_prev = l->lv_used_prev;
+  }
+
+  xfree(l);
+}
+
+/// Free a list, including all items it points to
+///
+/// Ignores the reference count. Does not do anything if
+/// tv_in_free_unref_items is true.
+///
+/// @param[in,out]  l  List to free.
+void tv_list_free(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  if (!tv_in_free_unref_items) {
+    tv_list_free_contents(l);
+    tv_list_free_list(l);
+  }
+}
+
+/// Unreference a list
+///
+/// Decrements the reference count and frees when it becomes zero or less.
+///
+/// @param[in,out]  l  List to unreference.
+void tv_list_unref(list_T *const l)
+{
+  if (l != NULL && --l->lv_refcount <= 0) {
+    tv_list_free(l);
+  }
+}
+
+//{{{2 Add/remove
+
+/// Remove items "item" to "item2" from list "l".
+///
+/// @warning Does not free the listitem or the value!
+///
+/// @param[out]  l  List to remove from.
+/// @param[in]  item  First item to remove.
+/// @param[in]  item2  Last item to remove.
+void tv_list_remove_items(list_T *const l, listitem_T *const item,
+                          listitem_T *const item2)
+  FUNC_ATTR_NONNULL_ALL
+{
+  // Notify watchers.
+  for (listitem_T *ip = item; ip != item2->li_next; ip = ip->li_next) {
+    l->lv_len--;
+    tv_list_watch_fix(l, ip);
+  }
+
+  if (item2->li_next == NULL) {
+    l->lv_last = item->li_prev;
+  } else {
+    item2->li_next->li_prev = item->li_prev;
+  }
+  if (item->li_prev == NULL) {
+    l->lv_first = item2->li_next;
+  } else {
+    item->li_prev->li_next = item2->li_next;
+  }
+  l->lv_idx_item = NULL;
+}
+
+/// Insert list item
+///
+/// @param[out]  l  List to insert to.
+/// @param[in,out]  ni  Item to insert.
+/// @param[in]  item  Item to insert before. If NULL, inserts at the end of the
```

Comment:
If `item` is the first element in another list, we would only have access to the old items of `l` with `l->lv_last`. `l->lv_first` would point to the list items of the other list.